File: apps/python-api/src/yourmedportal_api/application/public/public_report_service.py
from __future__ import annotations

from urllib.parse import quote

from yourmedportal_api.config import get_settings
from yourmedportal_api.lib.calendar import format_slot_from_db_to_iso
from yourmedportal_api.infrastructure.repositories.report_attachment_repository import ReportAttachmentRepository
from yourmedportal_api.infrastructure.repositories.report_repository import ReportRepository
from yourmedportal_api.infrastructure.repositories.visit_type_repository import VisitTypeRepository
from yourmedportal_api.rest.http_error import HttpError


class PublicReportService:
    def __init__(
        self,
        reports: ReportRepository,
        visit_types: VisitTypeRepository,
        attachments: ReportAttachmentRepository,
    ) -> None:
        self._reports = reports
        self._visit_types = visit_types
        self._attachments = attachments

    def get_report(self, fiscal_code: str, code: str) -> dict:
        normalized = code.strip().upper()
        rows = self._reports.find_by_code_and_fiscal(normalized, fiscal_code)
        if not rows:
            raise HttpError(404, "Referto non trovato")
        r = rows[0]
        vt_names = self._visit_types.select_name_by_id(r.visit_type_id)
        vtname = vt_names[0] if vt_names else "—"
        files = self._attachments.list_by_report_id(r.id) if r.status == "ready" else []
        base = get_settings().public_api_base()
        ex = r.exam_date
        exam_iso = format_slot_from_db_to_iso(ex) if hasattr(ex, "year") else str(ex)
        return {
            "status": r.status,
            "code": r.code,
            "examDate": exam_iso,
            "visitTypeName": vtname,
            "firstName": r.first_name,
            "lastName": r.last_name,
            "notes": r.notes,
            "attachments": [
                {
                    "id": f["id"],
                    "name": f["originalName"],
                    "mimeType": f["mimeType"],
                    "downloadUrl": (
                        f"{base}/public/attachments/{f['id']}"
                        f"?fiscalCode={quote(fiscal_code)}&reportCode={quote(normalized)}"
                    ),
                }
                for f in files
            ],
        }

File: apps/api/src/application/public/public-report-service.ts
import type { ReportAttachmentRepository } from '../../infrastructure/repositories/report-attachment-repository.js';
import type { ReportRepository } from '../../infrastructure/repositories/report-repository.js';
import type { VisitTypeRepository } from '../../infrastructure/repositories/visit-type-repository.js';
import { HttpError } from '../../rest/http-error.js';

export class PublicReportService {
	constructor(
		private readonly reports: ReportRepository,
		private readonly visitTypes: VisitTypeRepository,
		private readonly attachments: ReportAttachmentRepository
	) {}

	private get publicApiBase() {
		return process.env.PUBLIC_API_URL ?? `http://localhost:${process.env.API_PORT ?? '3001'}`;
	}

	async getReport(input: { fiscalCode: string; code: string }) {
		const normalizedCode = input.code.trim().toUpperCase();
		const [r] = await this.reports.findByCodeAndFiscal(normalizedCode, input.fiscalCode);
		if (!r) {
			throw new HttpError(404, 'Referto non trovato');
		}
		const [vt] = await this.visitTypes.selectNameById(r.visitTypeId);
		const files =
			r.status === 'ready' ? await this.attachments.listByReportId(r.id) : [];
		const base = this.publicApiBase;
		return {
			status: r.status,
			code: r.code,
			examDate: r.examDate.toISOString(),
			visitTypeName: vt?.name ?? '—',
			firstName: r.firstName,
			lastName: r.lastName,
			notes: r.notes,
			attachments: files.map((f) => ({
				id: f.id,
				name: f.originalName,
				mimeType: f.mimeType,
				downloadUrl: `${base}/public/attachments/${f.id}?fiscalCode=${encodeURIComponent(input.fiscalCode)}&reportCode=${encodeURIComponent(normalizedCode)}`
			}))
		};
	}
}
